NET-539: storage node assignment synchronization

harbu opened the pull request:
## Changes in a nutshell
- Add `blockNumber` to NodeRegistry's  `EtherumStorageEvent` and to the result of method `getStoredStreamsOf`, allowing us to determine which blockchain block the event or result applies to.
- Use this information in broker's storage plugin to synchronize events and polls appropriately. 
- Deprecate storage assignment event stream listening (we use smart contract event listeners to achieve the same) and thus broker's configuration setting `storageConfig.streamrAddress`.

## In detail

The client changes are straightforward. In the case of `EtherumStorageEvent`, we dig up the `blockNumber` directly from the ethers library callback function argument. For method `getStoredStreamsOf`, we modify the GraphQL query to also retrieve the latest block number with the result. In addition to these changes, some minor fixes to dependency injection to avoid circular dependency hassles that came up.

Broker has more major changes. Firstly, the core data structure and procedure for managing state is encapsulated in class `SetMembershipSynchronizer`. It is generalized to solve the problem of set membership over discrete time given full state updates and patches. In our case, the "elements" of the set are effectively `StreamPartID`(s) and we use The Graph polling for full state updates and blockchain-based events for patches.

Secondly, two separate data sources have been extracted from `StoragePlugin`. Everything related to polling full state data is in class `StoragePoller`, while event-based data handling can be found under class `StorageEventListener`. What responsibilities are left in `StoragePlugin` are (a) converting stream ids to stream parts, (b) taking care of storage cluster membership checks w.r.t. stream parts, and (c) instantiating the `StoragePoller` and `StorageEventListener` and connecting them with `SetMembershipSynchronizer`. Effectively `StoragePlugin` orchestrates the whole process of storage node assignments.

harbu commented on src/logic/resendStrategies.js:
```diff
@@ -176,8 +176,11 @@ class ProxiedResend {
         this.nodeToNode.send(neighborId, this.request).then(() => {
             this.currentNeighbor = neighborId
             this._resetTimeout()
+            return true
```
I don't understand why we need to return boolean here, is it used somewhere?

mirpo replied:
Add only to make happy eslint https://github.com/xjamundx/eslint-plugin-promise/blob/HEAD/docs/rules/always-return.md
Anyway, I think we'll add later strict rules (https://github.com/streamr-dev/eslint-config-streamr/pull/9), so it'll catch all possible errors. For now => reverted back.

harbu replied:
ok makes sense

harbu commented on src/logic/Node.js:
```diff
@@ -335,17 +335,21 @@ class Node extends EventEmitter {
             })
         }
+
+        return false
```
Don't understand need for `return false` and `return true`? Are they used somewhere?

harbu commented on test/unit/ResendHandler.test.js:
```diff
@@ -375,12 +375,13 @@ describe('ResendHandler', () => {
 
             let streamHasEnded = false
-            waitForStreamToEnd(resendHandler.handleRequest(request, 'source'))
+            await waitForStreamToEnd(resendHandler.handleRequest(request, 'source'))
```
No need for `await` since we use `promise.finally()` here that waits for promise to resolve/reject to call the callback. Actually I think we should not have `await` here for the test to be correct.

harbu commented on src/logic/Node.js:
```diff
@@ -335,17 +335,21 @@ class Node extends EventEmitter {
-                streamId,
-                node
+            // more strict, so when we get reject from connect, it will be caught
+            return this.protocols.nodeToNode.sendSubscribe(node, streamId).then(() => {
```
To me the code looks equivalent to before. But we get better error messages with this new version?

mirpo replied:
Better messages and better code!

harbu committed:
feat(client): add blockNumber to EthereumStorageEvent

Add `blockNumber` to EthereumStorageEvent so that it is possible to
determine whether an event is stale or not from the perspective of the
consumer of events.

harbu committed:
feat(client): add blockNumber to getStoredStreamsOf

Add `blockNumber` to NodeRegistry#getStoredStreamsOf. Turns out this was
relativly simple to do in The Graph.

harbu committed:
test(client): integration test for registerStorageEventListener

Add an integration test for registerStorageEventListener that verifies
that it picks up on storage node assignment and removal events.

harbu committed:
fix(client): fix DI circular dependency issues

harbu committed:
test(client): integration test for getStoredStreamsOf

harbu committed:
feat(client): StorageAssignmentSynchronizer

Implemented new class `StorageAssignmentSynchronizer` that ingests both
full-state polled assignment updates as well as event-based
partial-state ones. Based on the block numbers of these updates, it
combines both update types to construct a synchronized up-to-date state.

harbu committed:
feat(broker): StorageEventListener

Extract class `StorageEventListener` from `StorageConfig` that is
responsible for handling real-time storage node assignment events
learned about via streamr-client (which in turn learns about them via
smart contracts).

Removing the duplicate functionality from `StorageConfig` follows in a
later commit.

harbu committed:
feat(client): export EthereumStorageEvent

harbu committed:
feat(broker): StoragePoller

Polls full-state storage node assignments from The Graph via
StreamrClient#getStoredStreamsOf on a regular basis.

harbu committed:
test(broker): rm StoragePlugin.test.ts

Removed StoragePlugin test for the time being. We should think through
if we need a new test for this and how that test should work.

harbu committed:
feat(broker): wire up everything

Wire up StoragePlugin and StorageConfig to work with new classes and
functionality.

harbu committed:
fix(client): fix NodeEndpoints.test.ts

harbu committed:
refactor(broker): rename StorageAssignmentSynchronizer -> SetMemebershipSynchronizer

harbu committed:
refactor(broker): small fixes

harbu committed:
feat(broker): rm unused config "storageConfig.streamrAddress"

teogeb commented on packages/broker/src/plugins/storage/SetMembershipSynchronizer.ts:
```diff
@@ -0,0 +1,116 @@
+    /**
+     * Ingest a patch update
+     * @param elements the set of elements relevant to this patch
+     * @param operation indicates whether this is an assignment or a removal
```
```suggestion
     * @param operation indicates whether this is an addition or a removal
```

teogeb commented on packages/broker/src/plugins/storage/SetMembershipSynchronizer.ts:
```diff
@@ -0,0 +1,116 @@
+ *      2. Patch updates that add or remove a subset at a given time step.
+ */
+export class SetMembershipSynchronizer<E extends string> {
+    private lastState: ReadonlySet<E> = new Set<E>()
```
The `lastState` field is actually not needed.

harbu replied:
Removed.

teogeb commented on packages/broker/src/plugins/storage/StorageConfig.ts:
```diff
@@ -1,288 +1,108 @@
 export class StorageConfig {
+    private readonly listener: StorageConfigListener
+    private readonly synchronizer = new SetMembershipSynchronizer<StreamPartID>()
+    private readonly clusterId: string
```
The `clusterId` field is redundant, not needed outside of the constructor.

harbu replied:
Removed

teogeb commented on packages/broker/src/plugins/storage/SetMembershipSynchronizer.ts:
```diff
@@ -0,0 +1,116 @@
+     * @param elements the set of _all_ elements at this sequence number
+     * @param sequenceNo the sequence number for this state
+     */
+    ingestState(elements: Set<E>, sequenceNo: number): Diff<E> {
```
Rename to `ingestSnapshot`? Then we'd have three clearly different words: `patch`, `snapshot` and `state`.

harbu replied:
Good name. Updated.

teogeb commented on packages/broker/src/plugins/storage/StoragePoller.ts:
```diff
@@ -0,0 +1,63 @@
+    private readonly clusterId: string
+    private readonly pollInterval: number
+    private readonly streamrClient: StreamrClient
+    private readonly handleResult: (streams: Stream[], block: number) => void
```
Rename e.g. to `onNewSnapshot`?

harbu replied:
Renamed

teogeb commented on packages/broker/src/plugins/storage/StorageEventListener.ts:
```diff
@@ -0,0 +1,46 @@
+export class StorageEventListener {
+    private readonly clusterId: string
+    private readonly streamrClient: StreamrClient
+    private readonly handleEvent: (stream: Stream, type: 'added' | 'removed', block: number) => void
```
Rename to `onEvent`?

harbu replied:
Renamed

teogeb commented on packages/broker/src/plugins/storage/StoragePoller.ts:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    async start(): Promise<void> {
+        const schedulePoll = async () => {
```
Could use `scheduleAtInterval` helper method from `src/helpers/scheduler.ts`

harbu replied:
fixed

teogeb commented on packages/broker/test/unit/plugins/storage/StorageConfig.test.ts:
```diff
@@ -1,146 +1,178 @@
+
+describe(StorageConfig, () => {
+    let getStoredStreamsOf: jest.Mock<Promise<{ streams: Stream[], blockNumber: number }>, [nodeAddress: string]>
+    let emitStorageAssignmentEvent: ((event: EthereumStorageEvent) => any) | undefined
```
Rename to `storageEventListener` (as already in`StorageEventListener.test.ts`)

harbu replied:
fix

teogeb commented on packages/broker/test/unit/plugins/storage/StorageEventListener.test.ts:
```diff
@@ -0,0 +1,75 @@
+            async getStream() {
+                return {
+                    id: 'streamId',
+                    partitions: 0
```
The `partitions` should be > 0. 

(Maybe good to use a value > 1, as multi-partition situation is not the most typical situation and therefore worth checking. Not very important in this as we don't have any partition processing here.)

harbu replied:
Fixed

teogeb commented on packages/client/src/Encrypt.ts:
```diff
@@ -15,7 +15,7 @@ export default class PublisherEncryption implements Stoppable {
     constructor(
         private streamEndpoints: StreamEndpointsCached,
-        private keyExchange: PublisherKeyExchange,
+        @inject(delay(() => PublisherKeyExchange)) private keyExchange: PublisherKeyExchange,
```
Does this fix something related to this PR?

harbu replied:
Yeah the import paths have changed as a result of these changes, thereby causing circular dependencies.

teogeb reviewed (commented):
Great improvement 👍

harbu committed:
fix(client): NodeEndpoints.test.ts

harbu committed:
fix(broker): rm streamrAddress from test/utils.ts

harbu committed:
fix(broker): test/integration/broker.test.ts

Commit:
Update packages/broker/src/plugins/storage/SetMembershipSynchronizer.ts

Co-authored-by: Teo Gebhard <teo@streamr.com>

harbu committed:
fix(broker): rm lastState from SetMembershipSynchronizer

harbu committed:
refactor(broker): rm unused field from StorageConfig

harbu committed:
refactor(broker): rename state -> snapshot (when applies)

harbu committed:
refactor(broker): rename onHandleResult => onNewSnapshot

harbu committed:
refactor(broker): rename handleEvent => onEvent

harbu committed:
refactor(broker): use scheduleAtInterval in StoragePoller

harbu committed:
refactor(broker): StorageConfig.test.ts

harbu committed:
refactor(broker): fix partitions unrepresentative value in test

harbu committed:
fix(broker): rm test helper StorageAssignmentEventManager

harbu committed:
fix(broker): StorageNode.test.ts

harbu committed:
Merge branch 'main' into NET-539-storage-node-assignment-synchronization
